security/advancedtls/provider_test.go

```diff
@@ -0,0 +1,213 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package advancedtls
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"math/big"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/grpc/credentials/tls/certprovider"
+	"google.golang.org/grpc/security/advancedtls/testdata"
+)
+
+// The IdentityPemFileProvider is tested in different stages.
+// At stage 0, we copy the first set of certFile and keyFile to the temp files
+// that are watched by the goroutine.
+// At stage 1, we copy the second set of certFile and keyFile to the temp files
+// and verify the credential files are updated.
+func (s) TestIdentityPemFileProvider(t *testing.T) {
+	// Load certificates.
+	cs := &certStore{}
+	err := cs.loadCerts()
+	if err != nil {
+		t.Errorf("cs.loadCerts() failed: %v", err)
+	}
+	// Create temp files that are used to hold identity credentials.
+	certTmpPath, err := ioutil.TempFile(os.TempDir(), "pre-")
+	if err != nil {
+		t.Errorf("ioutil.TempFile(os.TempDir(), pre-) failed: %v", err)
+	}
+	defer os.Remove(certTmpPath.Name())
+	keyTmpPath, err := ioutil.TempFile(os.TempDir(), "pre-")
+	if err != nil {
+		t.Errorf("ioutil.TempFile(os.TempDir(), pre-) failed: %v", err)
+	}
+	defer os.Remove(keyTmpPath.Name())
+	tests := []struct {
+		desc           string
+		certFileBefore string
+		keyFileBefore  string
+		wantKmBefore   certprovider.KeyMaterial
+		certFileAfter  string
+		keyFileAfter   string
+		wantKmAfter    certprovider.KeyMaterial
+	}{
+		{
+			desc:           "Test identity provider on the client side",
+			certFileBefore: testdata.Path("client_cert_1.pem"),
+			keyFileBefore:  testdata.Path("client_key_1.pem"),
+			wantKmBefore:   certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}},
+			certFileAfter:  testdata.Path("client_cert_2.pem"),
+			keyFileAfter:   testdata.Path("client_key_2.pem"),
+			wantKmAfter:    certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer2}},
+		},
+		{
+			desc:           "Test identity provider on the server side",
+			certFileBefore: testdata.Path("server_cert_1.pem"),
+			keyFileBefore:  testdata.Path("server_key_1.pem"),
+			wantKmBefore:   certprovider.KeyMaterial{Certs: []tls.Certificate{cs.serverPeer1}},
+			certFileAfter:  testdata.Path("server_cert_2.pem"),
+			keyFileAfter:   testdata.Path("server_key_2.pem"),
+			wantKmAfter:    certprovider.KeyMaterial{Certs: []tls.Certificate{cs.serverPeer2}},
+		},
```

Comment:
Can we also have a test for case when there is a key-cert msimatch, and causing the failure of the updates?